### ISSUE TITLE
Write test suite

### ISSUE THREAD

**Issue by sandrotosi** [sandrotosi/puddletag#21]
Every python codebase benefits from a test suite; puddletag has lots of python3 code, but no unittests for it yet.

This is a great project for people that wish to learn how to write unittest (using pytest) for a large codebase, getting used to the process of testing code and this will help maintaining puddletag.

Feel free to comment here if you're interested in it, and i'll give out a few points where to start

**Comment by pedrobaeza** [comment]
Yeah, that was something I missed when first collaborating here, though I'm not sure what to test:

- Tags are written by third party lib (mutagen).
- Interface are not very prone for unit test.

**Comment by sandrotosi** [comment]
I'm making several low level changes lately, and i feel like i'm flying blind. Some functions rightfully so uses encode/decode when dealing with binary data, coming from files read directly, but _which ones_? also there are probably several functions no longer needed, but to be sure it'd be good to have coverage too.

So i'd start small, and from the main execution line: so starting from the gui and its main function: can i write a set of tests for:

* reading a directory
* parse tags (mocking the mp3 files)
* etc

and go from there

There's also some docs about writing unittest for a PyQt GUI [here](https://wiki.python.org/moin/PyQt/GUI_Testing) and [here](https://stackoverflow.com/questions/15044447/how-do-i-unit-testing-my-gui-program-with-python-and-pyqt) (from a very quick google search, have no experience with them)

**Comment by blueblots** [comment]
Is there any need for tests to be written for the code that isn't directly related to the GUI code? I don't have any experience with PyQt..

**Comment by sandrotosi** [comment]
> Is there any need for tests to be written for the code that isn't directly related to the GUI code? I don't have any experience with PyQt..

absolutely! my suggestion to start from the gui is to be sure to tests functions that are actually used (as i fear there is plenty of code no longer even reachable), but those functions dont necessarily have to be pyqt-related: like reading a mp3 file, and reading its tags, that's definitely used :)

**Comment by sandrotosi** [comment]
i dont think we should be limited by the platform necessarily.

simplicity and incremental steps are key here: let's start with a small pytest set of tests and we're gonna build on top of that.
